resolver.go

```diff
@@ -221,28 +235,58 @@ func (r *resolver) ServeDNS(w dns.ResponseWriter, query *dns.Msg) {
 		return
 	}
 
+	proto := w.LocalAddr().Network()
 	if resp == nil {
-		if len(r.extDNS) == 0 {
-			return
-		}
-
-		num := maxExtDNS
-		if len(r.extDNS) < maxExtDNS {
-			num = len(r.extDNS)
-		}
-		for i := 0; i < num; i++ {
-			log.Debugf("Querying ext dns %s:%s for %s[%d]", w.LocalAddr().Network(), r.extDNS[i], name, query.Question[0].Qtype)
+		for i := 0; i < maxExtDNS; i++ {
+			extDNS := &r.extDNSList[i]
+			if extDNS.ipStr == "" {
+				break
+			}
+			log.Debugf("Querying ext dns %s:%s for %s[%d]", proto, extDNS.ipStr, name, query.Question[0].Qtype)
+
+			extConn = extDNS.extConn
+			if extConn == nil || proto == "tcp" {
+				r.sb.execFunc(func() {
+					addr := fmt.Sprintf("%s:%d", extDNS.ipStr, 53)
+					extConn, err = net.DialTimeout(proto, addr, extIOTimeout)
+				})
+				if err != nil {
+					log.Debugf("Connect failed, %s", err)
+					continue
+				}
+				// For udp clients, persist the connection so that it can be reused
+				// for further queries
+				if proto == "udp" {
+					extDNS.extConn = extConn
+				}
+			}
 
-			c := &dns.Client{Net: w.LocalAddr().Network()}
-			addr := fmt.Sprintf("%s:%d", r.extDNS[i], 53)
+			// Timeout has to be set for every IO operation.
+			extConn.SetDeadline(time.Now().Add(extIOTimeout))
+			co := &dns.Conn{Conn: extConn}
+
+			err = co.WriteMsg(query)
+			if err != nil {
+				log.Debugf("Send to DNS server failed, %s", err)
+				if proto == "tcp" {
```

Comment:
Shouldn't we do this in a defer? So that we will know we always cleanup now and in future if more error conditions or other paths are introduced in the code path.